core/new-gui/src/app/workspace/service/workflow-graph/model/coeditor-presence.service.ts

```diff
@@ -0,0 +1,316 @@
+            userState.user.clientId !== this.texeraGraph.sharedModel.clientId
+        );
+        for (const clientId of change.added) {
+          const coeditorState = this.texeraGraph.sharedModel.awareness.getStates().get(clientId) as UserState;
+          if (coeditorState.user.clientId !== this.texeraGraph.sharedModel.clientId) this.addCoeditor(coeditorState);
+        }
+
+        for (const clientId of change.removed) {
+          if (!this.texeraGraph.sharedModel.awareness.getStates().has(clientId))
+            this.removeCoeditor(clientId.toString());
+        }
+
+        for (const clientId of change.updated) {
+          const coeditorState = this.texeraGraph.sharedModel.awareness.getStates().get(clientId) as UserState;
+          if (clientId.toString() !== this.texeraGraph.sharedModel.clientId) {
+            if (!this.hasCoeditor(clientId.toString())) {
+              this.addCoeditor(coeditorState);
+            } else {
+              this.updateCoeditorState(clientId.toString(), coeditorState);
+            }
+          }
+        }
+      }
+    );
+  }
+
+  /**
+   * Returns whether this co-editor is already recorded here.
+   * @param clientId
+   */
+  public hasCoeditor(clientId?: string) {
+    return this.coeditors.find(v => v.clientId === clientId);
+  }
+
+  /**
+   * Adds a new co-editor and initialize UI-updates for this editor.
+   * @param coeditorState
+   */
+  public addCoeditor(coeditorState: UserState) {
+    const coeditor = coeditorState.user;
+    if (!this.hasCoeditor(coeditor.clientId) && coeditor.clientId) {
+      this.coeditors.push(coeditor);
+      this.coeditorStates.set(coeditor.clientId, coeditorState);
+      this.updateCoeditorState(coeditor.clientId, coeditorState);
+    }
+  }
+
+  /**
+   * Removes a co-editor and clean up states recorded in this service.
+   * @param clientId
+   */
+  public removeCoeditor(clientId: string) {
+    for (let i = 0; i < this.coeditors.length; i++) {
+      const coeditor = this.coeditors[i];
+      if (coeditor.clientId === clientId) {
+        this.updateCoeditorState(clientId, {
+          user: coeditor,
+          userCursor: { x: 0, y: 0 },
+          currentlyEditing: undefined,
+          isActive: false,
+          highlighted: undefined,
+          changed: undefined,
+        });
+        this.coeditors.splice(i);
+      }
+    }
+    this.coeditorStates.delete(clientId);
+    if (this.shadowingModeEnabled && this.shadowingCoeditor?.clientId === clientId) {
+      this.stopShadowing();
+    }
+  }
+
+  /**
+   * Given a new <code>{@link UserState}</code> with specified clientId, this method updates this co-editor's presence
+   * information and corresponding UIs. This is an incremental update, i.e., it will first check existing states and
+   * only update what is new.
+   * @param clientId
+   * @param coeditorState
+   */
+  public updateCoeditorState(clientId: string, coeditorState: UserState) {
```

Comment:
this function is way too long, I suggest we split into at least 5-6 functions, one for each functionaility, and one function specific for shadowing related features

Reply:
Done.